src/cascadia/TerminalApp/TerminalWindow.cpp

```diff
@@ -0,0 +1,1262 @@
+                        return true;
+                    }
+                    // otherwise, keep walking. bubble the event manually.
+                }
+
+                if (auto focusedElement{ focusedObject.try_as<Windows::UI::Xaml::FrameworkElement>() })
+                {
+                    focusedObject = focusedElement.Parent();
+
+                    // Parent() seems to return null when the focusedElement is created from an ItemTemplate.
+                    // Use the VisualTreeHelper's GetParent as a fallback.
+                    if (!focusedObject)
+                    {
+                        focusedObject = winrt::Windows::UI::Xaml::Media::VisualTreeHelper::GetParent(focusedElement);
+                    }
+                }
+                else
+                {
+                    break; // we hit a non-FE object, stop bubbling.
+                }
+            } while (focusedObject);
+        }
+        return false;
+    }
+
+    // Method Description:
+    // - Used to tell the app that the 'X' button has been clicked and
+    //   the user wants to close the app. We kick off the close warning
+    //   experience.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TerminalWindow::CloseWindow(LaunchPosition pos)
+    {
+        if (_root)
+        {
+            // If persisted layout is enabled and we are the last window closing
+            // we should save our state.
+            if (_settings.GlobalSettings().ShouldUsePersistedLayout() && _numOpenWindows == 1)
+            {
+                if (const auto layout = _root->GetWindowLayout())
+                {
+                    layout.InitialPosition(pos);
+                    const auto state = ApplicationState::SharedInstance();
+                    state.PersistedWindowLayouts(winrt::single_threaded_vector<WindowLayout>({ layout }));
+                }
+            }
+
+            _root->CloseWindow(false);
+        }
+    }
+
+    winrt::TerminalApp::TaskbarState TerminalWindow::TaskbarState()
+    {
+        if (_root)
+        {
+            return _root->TaskbarState();
+        }
+        return {};
+    }
+
+    winrt::Windows::UI::Xaml::Media::Brush TerminalWindow::TitlebarBrush()
+    {
+        if (_root)
+        {
+            return _root->TitlebarBrush();
+        }
+        return { nullptr };
+    }
+    void TerminalWindow::WindowActivated(const bool activated)
+    {
+        if (_root)
+        {
+            _root->WindowActivated(activated);
+        }
+    }
+
+    // Method Description:
+    // - Returns true if we should exit the application before even starting the
+    //   window. We might want to do this if we're displaying an error message or
+    //   the version string, or if we want to open the settings file.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - true iff we should exit the application before even starting the window
+    bool TerminalWindow::ShouldExitEarly()

```

Comment:
this seems more like an AppLogic thing

Reply:
resolving as a part of the "commandline args are per-window" conversation